state/pruner.go

```diff
@@ -0,0 +1,267 @@
+	return &prunerConfig{
+		interval: config.DefaultPruningInterval,
+	}
+}
+
+type PrunerOption func(*prunerConfig)
+
+// PrunerInterval allows control over the interval between each run of the
+// pruner.
+func PrunerInterval(t time.Duration) PrunerOption {
+	return func(p *prunerConfig) { p.interval = t }
+}
+
+func NewPruner(stateStore Store, bs BlockStore, logger log.Logger, options ...PrunerOption) *Pruner {
+	cfg := defaultPrunerConfig()
+	for _, opt := range options {
+		opt(cfg)
+	}
+	p := &Pruner{
+		bs:         bs,
+		stateStore: stateStore,
+		logger:     logger,
+		interval:   cfg.interval,
+	}
+	p.BaseService = *service.NewBaseService(logger, "Pruner", p)
+	return p
+}
+
+func (p *Pruner) OnStart() error {
+	go p.pruningRoutine()
+	return nil
+}
+
+// SetApplicationRetainHeight sets the application retain height with some
+// basic checks on the requested height.
+//
+// If a higher retain height is already set, we cannot accept the requested
+// height because the blocks might have been pruned.
+//
+// If the data companion has already set a retain height to a higher value we
+// also cannot accept the requested height as the blocks might have been
+// pruned.
+func (p *Pruner) SetApplicationRetainHeight(height int64) error {
+	if height <= 0 || height < p.bs.Base() || height > p.bs.Height() {
+		return ErrInvalidHeightValue
+	}
+	currentAppRetainHeight, err := p.stateStore.GetApplicationRetainHeight()
+	if err != nil {
+		if err == ErrKeyNotFound {
+			currentAppRetainHeight = height
+		} else {
+			return err
+		}
+	}
+	currentCompanionRetainHeight, err := p.stateStore.GetCompanionBlockRetainHeight()
+	var noCompanionRetainHeight bool
+	if err != nil {
+		if err == ErrKeyNotFound {
+			noCompanionRetainHeight = true
+		} else {
+			return err
+		}
+	}
+	if currentAppRetainHeight > height || (!noCompanionRetainHeight && currentCompanionRetainHeight > height) {
+		return errors.New("cannot set a height lower than previously requested - blocks might have already been pruned")
+	}
+	err = p.stateStore.SaveApplicationRetainHeight(height)
+	return err
+}
+
+// SetCompanionRetainHeight sets the application retain height with some basic
+// checks on the requested height.
+//
+// If a higher retain height is already set, we cannot accept the requested
+// height because the blocks might have been pruned.
+//
+// If the application has already set a retain height to a higher value we also
+// cannot accept the requested height as the blocks might have been pruned.
+func (p *Pruner) SetCompanionRetainHeight(height int64) error {
+	if height <= 0 || height < p.bs.Base() || height > p.bs.Height() {
+		return ErrInvalidHeightValue
+	}
+	currentCompanionRetainHeight, err := p.stateStore.GetCompanionBlockRetainHeight()
+	if err != nil {
+		if err == ErrKeyNotFound {
+			currentCompanionRetainHeight = height
+		} else {
+			return err
+		}
+	}
+	currentAppRetainHeight, err := p.stateStore.GetApplicationRetainHeight()
+	var noAppRetainHeight bool
```

Comment:
Shouldn't this be initialized to false?

Reply:
Is it not initialized with the default value for bool (false)?